### DISCUSSION TITLE
Use custom_payload to generate matrix

### DISCUSSION THREAD

**Discussion by Kerizer** [community/community#25147] [Actions]
<p>We want to set matrix dynamicaly, based on some conditions. </p>
<p>This code:</p>
<pre><code>on:
  repository_dispatch

jobs:
  distribute-codepush-update:
    strategy:
      matrix:
        platform: ${{ github.event.client_payload.platform }}
        env: ${{ github.event.client_payload.env }}
    runs-on: ubuntu-latest
    name: Distribute ${{ matrix.platform }} ${{ matrix.env }} CodePush update
    timeout-minutes: 20
</code></pre>
<p>returns this error:<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/5/534a0a7a58bc0015717c0b217a53f5b5d4f904f9.png" data-download-href="/uploads/short-url/bSOgf9ySGZbE2CdO1flYRjJrlIB.png?dl=1" title="Screenshot 2020-03-23 at 11.08.21.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/21887713/181089144-d969edea-4d72-4874-9872-3e79030ecc8d.png" alt="Screenshot 2020-03-23 at 11.08.21.png" data-base62-sha1="bSOgf9ySGZbE2CdO1flYRjJrlIB" width="620" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/5/534a0a7a58bc0015717c0b217a53f5b5d4f904f9_2_620x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/5/534a0a7a58bc0015717c0b217a53f5b5d4f904f9_2_930x750.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/5/534a0a7a58bc0015717c0b217a53f5b5d4f904f9.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/5/534a0a7a58bc0015717c0b217a53f5b5d4f904f9_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screenshot 2020-03-23 at 11.08.21.png</span><span class="informations">1228×990 83.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>As we can see it is possible to set matrix variables using <code>github.event.client_payload.platform</code> and  it starts 4 jobs and inserts values in the &lt;job&gt;.name properly, but for some reason it is not starting.</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/kerizer">@kerizer</a> ,</p>
<p>I can repro the same on my side. Currently it’s not supported to set dynamic matrix value, you have to set the exact value for matrix. It’s answered in the similar <a href="https://github.community/t/how-to-create-dynamic-matrix-value/17326/2">ticket</a>.</p>
<p>Thanks.</p>

**Comment by Kerizer** [comment]
<p>But it’s clear that it starts 4 jobs and use correct values in <code>\&lt;job\&gt;.name</code>, so, whould it be possible to set them dynamically eventually? With current implementation we should either use matrix and fail unwanted jobs on first step or copy-paste same job 4 times (which is a bit hard to support, taking in considararation that we have 5 repos with same workflows and more are coming, so it’s 4*5).</p>

**Comment by chrispat** [comment]
<p>What does your payload look like?  Are you trying to pass a delimited string or is it a proper JSON array? </p>

**Comment by chrispat** [comment]
<p>Here is an example that works for me.</p>
<p>Workflow that triggers the repositor_dispatch</p>
<pre><code># This is a basic workflow to help you get started with Actions

name: repo-dispatch

# Controls when the action will run. Triggers the workflow on push or pull request
# events but only for the master branch
on:
  push:
    branches: [master]
  pull_request:
    branches: [master]

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  # This workflow contains a single job called "build"
  trigger:
    # The type of runner that the job will run on
    runs-on: ubuntu-latest

    # Steps represent a sequence of tasks that will be executed as part of the job
    steps:
      - name: Repository Dispatch
        uses: peter-evans/repository-dispatch@v1
        with:
          token: ${{ secrets.REPO_ACCESS_TOKEN }}
          event-type: 'matrix-ci'
          client-payload: '{"images": ["ubuntu-latest", "windows-latest"] }'
</code></pre>
<p>Workflow that recieves it</p>
<pre><code>name: .NET Core. CI 1

on:
  push:
    branches: [master]
  pull_request:
    branches: [master]
  repository_dispatch:
    types: [matrix-ci]

jobs:
  build:
    if: github.event != 'repository_dispatch'
    runs-on: ${{ matrix.image }}
    strategy:
      matrix:
        image: ['ubuntu-latest', 'windows-latest', 'macos-latest']

    steps:
    - uses: actions/checkout@v2
    - name: Setup .NET Core
      uses: actions/setup-dotnet@v1
      with:
        dotnet-version: 3.1.101
    - name: Build with dotnet
      run: dotnet build --configuration Release

  build-2:
    if: github.event == 'repository_dispatch'
    runs-on: ${{ matrix.image }}
    strategy:
      matrix:
        image: ${{ github.event.client_payload.images }}

    steps:
    - uses: actions/checkout@v2
    - name: Setup .NET Core
      uses: actions/setup-dotnet@v1
      with:
        dotnet-version: 3.1.101
    - name: Build with dotnet
      run: dotnet build --configuration Release
</code></pre>
<p>Successfule run</p>
<p><a href="https://github.com/chrispat/multi-ci-test/actions/runs/68415548" rel="nofollow noopener">https://github.com/chrispat/multi-ci-test/actions/runs/68415548</a></p>

**Comment by Bo98** [comment]
<p>Unless I’m reading it wrong, it looks like that build was run on build rather than build-2. It also ran on macos-latest which your trigger job does not specify.</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/kerizer">@kerizer</a> <a class="mention" href="/bo98">@bo98</a> ,</p>
<p>Now it looks as fixed. Please check my workflow: <a href="https://github.com/weide-zhou/ticket17/actions/runs/69528670" rel="nofollow noopener">https://github.com/weide-zhou/ticket17/actions/runs/69528670</a></p>
<p>Please check your workflow again, thanks.</p>

**Comment by chrispat** [comment]
<p>I cleaned up the example a little so I don’t have so many duplicate things on the same commit.</p>
<p>This one clearly ran only two jobs <a href="https://github.com/chrispat/multi-ci-test/actions/runs/69926570" rel="nofollow noopener">https://github.com/chrispat/multi-ci-test/actions/runs/69926570</a>.</p>

**Comment by Kerizer** [comment]
<p>That looks awesome! Thank you :slight_smile:</p>
